_dev/interop_test_runner.py

```diff
@@ -27,27 +27,47 @@ class Docker(object):
     def __init__(self):
         self.d = docker.from_env()
 
+    def close(self):
+        self.d.close()
+
     def get_ip(self, server):
         tris_localserver_container = self.d.containers.get(server)
         return tris_localserver_container.attrs['NetworkSettings']['IPAddress']
 
     def run_client(self, image_name, cmd):
         ''' Runs client and returns tuple (status_code, logs) '''
-        c = self.d.containers.create(image=image_name, command=cmd)
-        c.start()
+        c = self.d.containers.run(image=image_name, detach=True, command=cmd)
         res = c.wait()
-        ret = c.logs()
+        ret = c.logs().decode('utf8')
         c.remove()
         return (res['StatusCode'], ret)
 
     def run_server(self, image_name, cmd=None, ports=None, entrypoint=None):
         ''' Starts server and returns docker container '''
-        c = self.d.containers.create(image=image_name, detach=True, command=cmd, ports=ports, entrypoint=entrypoint)
-        c.start()
-        # TODO: maybe can be done better?
-        time.sleep(3)
+        c = self.d.containers.run(image=image_name, auto_remove=True, detach=True, command=cmd, ports=ports, entrypoint=entrypoint)
+        self.wait_for_ports(c)
         return c
 
+    def wait_for_ports(self, c):
```

Comment:
can we do it better without calling shell at all?

Reply:
I couldn't really find a good alternative. I tried to retrieve container IP and try to port scan it from Python. Does not work on Docker for Mac, [details](https://docs.docker.com/docker-for-mac/networking/#i-cannot-ping-my-containers).

An alternative is to start a container for the sole purpose of port-scanning a host (it could consist of a single Go binary), but the size of that support code will be slightly larger than the current approach. If you think that this would be better, I can change it though?

Reply:
I would double check docker-py. I wonder if it's not done by this wrapper itself.
If not can't we just "import socket" and "socket.connect"? It's hard for me to  believe python doesn't have something equivalent to ``ss``

Reply:
I have tried the socket approach before, but it won't work:
> Does not work on Docker for Mac, [details](https://docs.docker.com/docker-for-mac/networking/#i-cannot-ping-my-containers).

Checking for reachable ports is not something that can be done with the high-level docker-py API (https://github.com/docker/docker-py/issues/2229). Usually healthchecks are used to test whether a service is ready (e.g. aside from checking that a TCP socket can be opened, verify that the HTTP service returns a proper response). In our case, it is sufficient if a TCP socket can be opened.

Reply:
ok, keep digging.